### DISCUSSION TITLE
"latest release"

### DISCUSSION THREAD

**Discussion by MikeSpreitzer** [kubestellar/kubestellar#2218] [General]
There has been some dissatisfaction with how we talk about releases and whether or how to use the term "latest release" and terms like it. Let me summarize the agreement that I thought that we had, and then expand on the issue.

We have been using the word "release" to refer to anything that GitHub lists in the "releases" list for a repo. This includes test releases named according to the pattern "major.minor.patch-rcNum". This includes alpha releases like "0.23.0-alpha.2". This includes releases that are completely ordinary and regular, named according to the pattern "major.minor.patch" (e.g., "0.23.0"). We have been using the adjective "regular" to distinguish that latter category of releases from the others. In most English prose, we use the word "latest" to mean what you find in the dictionary --- _most recent_. We have never (since we last discussed terminology, maybe ever) made a patch release to an older major.minor after creating a higher major.minor; the question of what "latest" means in that state has not arisen.

For example: when the most recent release of any kind is "0.23.0-rc2" and the most recent non-test non-alpha release is "0.22.0", in English prose "the latest release" resolves to "0.23.0-rc2" and "the latest regular release" resolves to "0.22.0". This has the virtue that we do not need to re-educate people about the meanings of common English words and they combine in the usual ways. Some people have objected that it is not obvious enough what the adjective "regular" is trying to say.

One confounding factor is that there are a variety of relevant formal things that use the string "latest" in a special way. Following are some of them.

- As a tag in a container image repository. We do not tell anyone to pull images with the "latest" tag. Using the image tag "latest" is generally discouraged.
- As one of the aliases maintained by `mike` in our documentation (see `docs/README.md`). This is essentially invisible to users: we do not direct users to this alias. Contributors generally need to pay no heed to it; it is part of the doc machinery that exists and just works and contributors rarely mess with. Our doc machinery maintains the relationship that the alias named "latest" refers to the doc version that has the highest semver among those that start with "release-". Semantically, what we are trying to do with this alias is keep track of the version that we want to direct new users to; this will not be a test release or an alpha release. Technically, we use this alias for two things: (1) to identify which version should be seen by a user who enters the website without specifying a version, and (2) to identify which version does NOT include the version warning banner discussed below.
- GitHub has a list of releases and calls one of them "latest". As far as GitHub is concerned, it is up to the repo authors to choose which of the releases is the one that GitHub calls "latest". Thus, as far as GitHub mechanics are concerned, the meaning of this "latest" is up to each repo's authors. We choose to make GitHub's "latest" point to the most recent (or is it highest semver? I am not sure) regular release.
- Digression: GitHub allows a release to be marked as a "pre-release". This does NOT cause that release to be removed from the list of releases. Thus, a pre-release is a release. This marking is under the control of the repo authors. We choose to make candidate releases and alphas be known to GitHub as pre-releases. GitHub does not allow a pre-release to be marked as "latest".
- `mike` puts up a version warning banner on website pages that are not from a certain version (or two). The relevant version(s) are identified in the `mike` configuration (in `docs/mkdocs.yml`). We choose to make the version warning banner appear on pages in all versions except the one that the `mike` alias "latest" refers to.
- The wording in the warning banner is content that we author in an HTML template file. Our banner wording warns the reader that they are not looking at "the latest release" but the banner appears on pages for every version that is not the latest _regular_ release. So the wording of this warning is currently buggy.
- PR #2206 introduces doc variables (i.e., things defined in `docs/mkdocs.yml`) named `latest_release` and `latest_regular_release` --- with meanings consistent with the ordinary English meanings of those phrases.

**Comment by KPRoche** [comment]
> * The wording in the warning banner is content that we author in an HTML template file. Our banner wording warns the reader that they are not looking at "the latest release" but the banner appears on pages for every version that is not the latest _regular_ release. So the wording of this warning is currently buggy.

FYI: The message in the template for the new website design (from PR #2229 ) was emended to "latest regular release"
